### ISSUE TITLE
Add a `poetry check --sync` command

### ISSUE THREAD

**Issue by ValentinCalomme** [python-poetry/poetry#8846]
## Summary

In my development process, I often face version mismatches amongst developers, primarily due to updates in the `poetry.lock` file that do not violate the requirements present in `pyproject.toml`. This leads to issues such as linting failures from updated package stubs. While `poetry check --lock` ensures consistency between `pyproject.toml` and `poetry.lock`, it doesn't verify if the working environment matches the versions in `poetry.lock`.

## Proposal

I propose the addition of a new command: `poetry check --sync`. This command would check that the environment's package versions are in sync with those in `poetry.lock`, without performing any installations or updates.

## Justification

- Prevent Mismatches: This command would serve as a preventive measure against version mismatches, enhancing the reliability of our development environment.
- Adding this to the `poetry check` command group would also mean that it would be available in `pre-commit`
- Convenience: It provides a straightforward way to verify environment consistency without the side effects of installing or updating packages, unlike the current `poetry install --sync`.
- Filling a Gap: Currently, there is no option to simply check for version consistency. The `--dry-run` flag does not offer this functionality, indicating a gap that `poetry check --sync` could fill.

**Comment by dimbleby** [comment]
> the --dry-run flag does not offer this functionality

yes it does?  `poetry install --sync --dry-run` works as expected?

**Comment by ValentinCalomme** [comment]
> > the --dry-run flag does not offer this functionality
> 
> yes it does?  `poetry install --sync --dry-run` works as expected?

It does! What I am saying is that it doesn't offer the functionality I am describing.

Let's say I have a package with version requirement `^1.0.0` in my `pyproject.toml`. One of my team members updates the dependencies, and now the `poetry.lock` has version `1.0.1`.

The lock file is still valid based on the dependency requirements, so `poetry check --lock` will pass as it should.

If I run `poetry install --sync`, my package will update, as it should.

If I run `poetry install --sync --dry-run`, I won't get an error because the lock file is consistent with the requirements. To know if I do have the proper installation, I would need to parse the output of the command.

Which is why I am suggesting having a command that essentially tells me "your environment is consistent with your lock file"

**Comment by abn** [comment]
I am not sure there is much value added here for the maintenance burden

Can you clarify why you would not want to simply do `poetry install --sync`? The "side-effect of installing packages" is something you will want to do even after you check since you want to avoid the mismatch etc.

**Comment by BobbyRyterski** [comment]
I am similarly looking for this behavior. It's useful to tease apart detection and taking an action. The value is an explicit exit code which can be consumed elsewhere. Say the pre-commit as mentioned before. Or during CI, where a quick check is desired to support caching decisions and / or before a much longer install process.

For prior art, https://bundler.io/v2.5/man/bundle-check.1.html

**Comment by dsayling** [comment]
I am also looking for this behavior.

I think a simple approach could be to change value of the exit code when using `--dry-run`, i.e. 0 implies nothing needs to change, 1 something needs to be removed, 2 something needs to be upgraded/downgraded, 3 both

**Comment by dsayling** [comment]
for what it's worth, I'm currently just doing this to see if `poetry install --sync` would result in environment changes.

```console
poetry install --sync --dry-run 2>&1 | grep -q "0 installs, 0 updates, 0 removals"
```

Which you can take the exit code from that do what you will

```console
poetry install --sync --dry-run 2>&1 | grep -q "0 installs, 0 updates, 0 removals" || echo "🚨 Your venv is out of date"
```

**Comment by ofer-pd** [comment]
It's obvious to me that this functionality is needed.  It's an extremely common occurrence to have your `poetry.lock` file suddenly change under you as a result of running `git pull` and not notice that your venv is now out of sync with the latest lock file.  There needs to be a simple and reliable way to check for this (such as non-zero exit code on a poetry command) so tooling can be built on it.

@dsayling - I appreciate the one-liner, and I'll probably use it for now (what choice do I have), but that shouldn't be considered an acceptable substitute for the proper solution, which is a poetry command that returns a non-zero exit code.

**Comment by dsayling** [comment]
I agree that the functionality would be useful. 

I have a feeling it wouldn't be too much to implement either, but a maintainer should probably chime in. Maybe worth just implementing and just providing a PR.

**Comment by d-chris** [comment]
currently im developing a [plugin](https://github.com/d-chris/poetry-plugin-latest) to extend poetry commands with non-zero return values which i can use as pre-commit-hook.

i wanted a pre-commit-hook which a can execute on pre-push to check if all my dependencies are up-to-date. so wrapped `poetry show -o -T` into a new command.

@valentincalomme maybe this fits your requirement to, its still in development but soon i will release it to pypi

```cmd
poetry latest --only=main
```
